packages/xs-vat-worker/src/vatWorker.js

```diff
@@ -0,0 +1,196 @@
+}
+
+// see also: detecting an empty vat promise queue (end of "crank")
+// https://github.com/Agoric/agoric-sdk/issues/45
+function waitUntilQuiescent(setImmediate) {
+  return new Promise((resolve, _reject) => {
+    setImmediate(() => {
+      // console.log('hello from setImmediate callback. The promise queue is presumably empty.');
+      resolve();
+    });
+  });
+}
+
+function runAndWait(f, errmsg, setImmediate) {
+  Promise.resolve()
+    .then(f)
+    .then(undefined, err => workerLog(`doProcess: ${errmsg}:`, err));
+  return waitUntilQuiescent(setImmediate);
+}
+
+function makeWorker(io, setImmediate) {
+  let dispatch;
+
+  async function doProcess(dispatchRecord, errmsg) {
+    const dispatchOp = dispatchRecord[0];
+    const dispatchArgs = dispatchRecord.slice(1);
+    workerLog(`runAndWait`);
+    await runAndWait(
+      () => dispatch[dispatchOp](...dispatchArgs),
+      errmsg,
+      setImmediate,
+    );
+    workerLog(`doProcess done`);
+  }
+
+  function doNotify(vpid, vp) {
+    const errmsg = `vat.promise[${vpid}] ${vp.state} failed`;
+    switch (vp.state) {
+      case 'fulfilledToPresence':
+        return doProcess(['notifyFulfillToPresence', vpid, vp.slot], errmsg);
+      case 'redirected':
+        throw new Error('not implemented yet');
+      case 'fulfilledToData':
+        return doProcess(['notifyFulfillToData', vpid, vp.data], errmsg);
+      case 'rejected':
+        return doProcess(['notifyReject', vpid, vp.data], errmsg);
+      default:
+        throw Error(`unknown promise state '${vp.state}'`);
+    }
+  }
+
+  function sendUplink(msg) {
+    assert(msg instanceof Array, `msg must be an Array`);
+    io.writeMessage(JSON.stringify(msg));
+  }
+
+  // fromParent.on('data', data => {
+  //  workerLog('data from parent', data);
+  //  toParent.write('child ack');
+  // });
+
+  let syscallLog;
+  const handle = harden(async ([type, ...margs]) => {
+    workerLog(`received`, type);
+    if (type === 'start') {
+      // TODO: parent should send ['start', vatID]
+      workerLog(`got start`);
+      sendUplink(['gotStart']);
+    } else if (type === 'setBundle') {
+      const [bundle, vatParameters] = margs;
+      const endowments = {
+        console: makeConsole(`SwingSet:vatWorker`),
+        HandledPromise,
+      };
+      // ISSUE: this draft code is contorted because it started
+      // as code that didn't return anything but now it
+      // has to return a promise to be resolved before
+      // reading the next input.
+      return importBundle(bundle, { endowments }).then(vatNS => {
+        workerLog(`got vatNS:`, Object.keys(vatNS).join(','));
+        sendUplink(['gotBundle']);
+
+        function doSyscall(vatSyscallObject) {
+          sendUplink(['syscall', ...vatSyscallObject]);
+        }
+        const syscall = harden({
+          send: (...args) => doSyscall(['send', ...args]),
+          callNow: (..._args) => {
+            throw Error(`nodeWorker cannot syscall.callNow`);
```

Comment:
we should be able to get this one working, because we're using a blocking-mode socket, right? we don't need that to land this PR, but it's a significant feature that we need to have shortly thereafter.

(Node.js `Worker` threads cannot block, but that should only be a limitation of the other worker type, not this one)

Reply:
OK; this is a result of a wholesale copy-and-paste, so I didn't even notice that limitation.

Reply:
postponed as https://github.com/Agoric/agoric-sdk/issues/1617